master/internal/core_checkpoint.go

```diff
@@ -0,0 +1,397 @@
+package internal
+
+import (
+	"archive/tar"
+	"archive/zip"
+	"compress/gzip"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"github.com/google/uuid"
+	"github.com/labstack/echo/v4"
+	"github.com/pkg/errors"
+
+	"github.com/determined-ai/determined/master/internal/api"
+	"github.com/determined-ai/determined/master/pkg/ptrs"
+	"github.com/determined-ai/determined/master/pkg/schemas/expconf"
+)
+
+type archiveWriter interface {
+	WriteFileHeader(fname string, size int64) error
+	Write(b []byte) (int, error)
+}
+
+type tarArchiveWriter struct {
+	tw *tar.Writer
+}
+
+func (aw *tarArchiveWriter) WriteFileHeader(fname string, size int64) error {
+	hdr := tar.Header{
+		Name: fname,
+		Mode: 0666,
+		Size: size,
+	}
+	if strings.HasSuffix(fname, "/") {
+		// This a directory
+		hdr.Mode = 0777
+	}
+	return aw.tw.WriteHeader(&hdr)
+}
+
+func (aw *tarArchiveWriter) Write(p []byte) (int, error) {
+	return aw.tw.Write(p)
+}
+
+type zipArchiveWriter struct {
+	zw        *zip.Writer
+	zwContent io.Writer
+}
+
+func (aw *zipArchiveWriter) WriteFileHeader(fname string, size int64) error {
+	// Zip by default sets mode 0666 and 077 for files and folders respectively
+	zwc, err := aw.zw.Create(fname)
+	if err != nil {
+		return err
+	}
+	aw.zwContent = zwc
+	return nil
+}
+
+func (aw *zipArchiveWriter) Write(p []byte) (int, error) {
+	var w io.Writer
+	if aw.zwContent == nil {
+		return 0, nil
+	}
+	w = aw.zwContent
+	return w.Write(p)
+}
+
+type delayWriter struct {
+	delayBytes int
+	buf        []byte
+	next       io.Writer
+}
+
+func (w *delayWriter) Write(p []byte) (int, error) {
+	if w.buf == nil {
+		return w.next.Write(p)
+	}
+
+	w.buf = append(w.buf, p...)
+	if len(w.buf) < w.delayBytes {
+		return len(p), nil
+	}
+
+	n, err := w.next.Write(w.buf)
+	w.buf = nil
+	return n, err
+}
+
+// Flush the buffer if it is nonempty.
+func (w *delayWriter) Close() error {
+	if w.buf != nil && len(w.buf) > 0 {
+		_, err := w.next.Write(w.buf)
+		return err
+	}
+	return nil
+}
+
+func newDelayWriter(w io.Writer, delayBytes int) *delayWriter {
+	return &delayWriter{
+		delayBytes: delayBytes,
+		buf:        make([]byte, 0, delayBytes),
+		next:       w,
+	}
+}
+
+// S3SeqWriterAt satisfies S3 APIs' io.WriterAt interface while staying sequential.
+// Ref: https://docs.aws.amazon.com/sdk-for-go/api/service/s3/s3manager/#Downloader
```

Comment:
nit: can you explicitly mention that this only works when concurrent downloads is set to 1?  Otherwise an unsuspecting dev could think that s3SeqWriterAt is capable of a lot more than it is.

I think the comment where you set concurrent downloads is good, but I think you should have the comment in both places, since it's basically an implicit requirement that developers must follow for things to work right.

Reply:
Makes sense!